snap/validate.go

```diff
@@ -1188,15 +1188,114 @@ func NeededDefaultProviders(info *Info) (providerSnapsToContentTag map[string][]
 	return (SimplePrereqTracker{}).DefaultProviderContentAttrs(info, nil)
 }
 
-// ValidateBasesAndProviders checks that all bases/default-providers are part of the seed
-func ValidateBasesAndProviders(snapInfos []*Info) []error {
-	all := naming.NewSnapSet(nil)
-	for _, info := range snapInfos {
-		all.Add(info)
+// SelfContainedPrereqTracker is a stateful helper to track
+// prerequisites of snaps (default-providers in particular).
+// It is meant to be used when dealing with a self-contained set
+// of snaps, with no desire to fetch further snaps, so all
+// prerequisites must be present in the set itself.
+// This applies to first boot seeding and remodeling for example.
+// SelfContainedPrereqTracker implements snapstate.PrereqTracker.
+type SelfContainedSetPrereqTracker struct {
+	snaps []*Info
+	all   *naming.SnapSet
+}
+
+// NewSelfContainedSetPrereqTracker returns a new SelfContainedSetPrereqTracker.
+func NewSelfContainedSetPrereqTracker() *SelfContainedSetPrereqTracker {
+	return &SelfContainedSetPrereqTracker{
+		all: naming.NewSnapSet(nil),
 	}
+}
 
-	var errs []error
-	for _, info := range snapInfos {
+// Add adds a snap to track. Add implements snapstate.PrereqTracker.
+func (prqt *SelfContainedSetPrereqTracker) Add(info *Info) {
+	if !prqt.all.Contains(info) {
+		prqt.all.Add(info)
+		prqt.snaps = append(prqt.snaps, info)
+	}
+}
+
+// DefaultProviderContentAttrs implements snapstate.PrereqTracker.
+// Given how snapstate uses this and as SelfContainedSetPrereqTracker is for
+// when no automatic fetching of prerequisites is desired, this always returns
+// nil.
+func (prqt *SelfContainedSetPrereqTracker) DefaultProviderContentAttrs(info *Info, repo InterfaceRepo) map[string][]string {
```

Comment:
Could you explain the reasoning for returning `nil` here more? From looking at the doc comment on `DefaultProviderContentAttrs` in the `PrereqTracker` interface, I don't understand how other implementations besides the one in `SimplePrereqTracker` make sense.

Reply:
see the detail I added to https://github.com/snapcore/snapd/pull/13340/files#diff-9e832b3649d23b262a0cc4837c71c91a0e5d72ed1f2977cd7151b5a8e27da360R1097 

But I agree that maybe that comment needs to be changed further and the method needs a different name (the latter is a bit disruptive though at this point)

Reply:
I see, that makes sense. Thanks!

Reply:
ah, it is used indirectly in most of snapstate. So maybe renaming it to MissingProviderContentTags ? and changing a bit the doc comment in the interface?

Reply:
I think that would help, yes.